service/history/archivalQueueTaskExecutor.go

```diff
@@ -0,0 +1,162 @@
+	case *tasks.ArchiveExecutionTask:
+		err = e.processArchiveExecutionTask(ctx, task)
+	default:
+		err = fmt.Errorf("task with invalid type sent to archival queue: %+v", task)
+	}
+	return tags, true, err
+}
+
+func (e *archivalQueueTaskExecutor) processArchiveExecutionTask(ctx context.Context,
+	task *tasks.ArchiveExecutionTask) (err error) {
+	weContext, release, err := getWorkflowExecutionContextForTask(ctx, e.workflowCache, task)
+	if err != nil {
+		return err
+	}
+	defer func() { release(err) }()
+
+	mutableState, err := weContext.LoadMutableState(ctx)
+	if err != nil {
+		return err
+	}
+	if mutableState == nil || mutableState.IsWorkflowExecutionRunning() {
+		return errors.New("cannot archive workflow which is running")
+	}
+	branchToken, err := mutableState.GetCurrentBranchToken()
+	if err != nil {
+		return err
+	}
+
+	namespaceEntry := mutableState.GetNamespaceEntry()
+	namespaceName := namespaceEntry.Name()
+	nextEventID := mutableState.GetNextEventID()
+	closeFailoverVersion, err := mutableState.GetLastWriteVersion()
+	if err != nil {
+		return err
+	}
+
+	executionInfo := mutableState.GetExecutionInfo()
+	workflowTypeName := executionInfo.GetWorkflowTypeName()
+	startTime := executionInfo.GetStartTime()
+	if startTime == nil {
+		return errors.New("can't archive workflow with nil start time")
+	}
+	executionTime := executionInfo.GetExecutionTime()
+	if executionTime == nil {
+		return errors.New("can't archive workflow with nil execution time")
+	}
+	closeTime := executionInfo.GetCloseTime()
+	if closeTime == nil {
+		return errors.New("can't archive workflow with nil close time")
+	}
+	executionState := mutableState.GetExecutionState()
+	memo := getWorkflowMemo(copyMemo(executionInfo.Memo))
+	_, err = e.archiver.Archive(ctx, &archival.Request{
+		ShardID:              e.shardContext.GetShardID(),
+		NamespaceID:          task.NamespaceID,
+		Namespace:            namespaceName.String(),
+		WorkflowID:           task.WorkflowID,
+		RunID:                task.RunID,
+		BranchToken:          branchToken,
+		NextEventID:          nextEventID,
+		CloseFailoverVersion: closeFailoverVersion,
+		HistoryURI:           namespaceEntry.HistoryArchivalState().URI,
+		WorkflowTypeName:     workflowTypeName,
+		StartTime:            *startTime,
+		ExecutionTime:        *executionTime,
+		CloseTime:            *closeTime,
+		Status:               executionState.Status,
+		HistoryLength:        nextEventID - 1,
+		Memo:                 memo,
+		SearchAttributes:     getSearchAttributes(copySearchAttributes(executionInfo.SearchAttributes)),
+		VisibilityURI:        namespaceEntry.VisibilityArchivalState().URI,
+		Targets:              []archival.Target{archival.TargetHistory, archival.TargetVisibility},
+		CallerService:        common2.HistoryServiceName,
+	})
+	if err != nil {
+		return err
+	}
+	retention := namespaceEntry.Retention()
+	if retention == 0 {
+		retention = 7 * 24 * time.Hour
+	}
+	deleteTime := closeTime.Add(retention)
```

Comment:
Add a jitter for this

Reply:
Done in https://github.com/temporalio/temporal/pull/3565